hclsyntax: conditional refinements of collections must use Range()

jbardin opened the pull request:
Update cty to account for fixes in refinements.

When attempting to determine the final length range for a conditional expression with collections, the length values may still be unknown. Always use `Range()` to get the lower and upper bounds. Calls to `Range()` must also always be made with unmarked values.

apparentlymart reviewed (approved):
This seems plausible but given our earlier experience with a different part of the same function last week I'd suggest also making sure there's a variant of this test case where the true and false values are both marked, to make sure it handles that combination of marks and refinements correctly.

jbardin commented:
Yes, I'll probably add an additional test if I don't come back with another PR. I'm tracking down some refinements+marks oddities right now.

jbardin commented:
OK, the upstream problems I was tracking are taken care of in the latest cty, so I updated the version to match, and extended this PR to cover the marked cases.

apparentlymart reviewed (approved):
This looks good to me!

One small hazard here is that IIRC we'd left HCL's dependency to cty on v1.13 to avoid forcing callers to adopt the Unicode 15 variant of grapheme cluster segmentation, though it's probably been long enough since Go 1.21 came out for that to not be as much of a concern anymore. (The concern was pretty narrow anyway: it only applies to those who care about Unicode consistency between different components _and_ who need to remain on an older version of Go for some reason.)

Therefore I think this is probably fine though maybe we should make this be an HCL minor release (rather than a patch) to help communicate that upgrading could have some minor behavior changes beyond just fixing this bug.

apparentlymart commented on hclsyntax/expression_test.go:
```diff
@@ -1957,6 +1957,34 @@ EOT
+						CollectionLengthLowerBound(3).CollectionLengthUpperBound(4).NewValue(),
+				},
+			},
+			cty.UnknownVal(cty.Set(cty.String)).Refine().NotNull().CollectionLengthLowerBound(1).CollectionLengthUpperBound(4).NewValue().Mark("test"), // deduced through refinements
```
Post-approval little followup:

Where did the `NotNull` refinement come from here? It doesn't seem like either of the input values are refined as being not-null.

jbardin commented:
Now with more refined refinements! (and dropped the cty upgrade, since the tests no longer directly depend on it to pass)
Re-worked the range and null handling for number and collection types, and added a bunch more tests which failed before the fixes.

apparentlymart reviewed (approved):
Ahh, this version seems much nicer!

I guess when I was originally working on this I hadn't completed the `Value.Range` API in `cty` yet, since I don't know why else I would've manually implemented all of that finagling with the lower-level API. Using the range API throughout seems to make this much easier to follow.

jbardin commented on hclsyntax/expression.go:
```diff
@@ -696,64 +696,95 @@ func (e *ConditionalExpr) Value(ctx *hcl.EvalContext) (cty.Value, hcl.Diagnostic
+		falseResult, falseResultMarks := falseResult.Unmark()
+
+		// use a value to merge marks
+		_, resMarks := cty.DynamicVal.WithMarks(condResultMarks, trueResultMarks, falseResultMarks).Unmark()
```
Hehe, I thought that would get noticed ;) As soon as I started looking at conditional handling I wanted set operations on `ValueMarks`; maybe later.
